lark/parsers/lalr_analysis.py

```diff
@@ -31,6 +32,128 @@ def __init__(self, states, start_state, end_state):
         self.start_state = start_state
         self.end_state = end_state
 
+
+class ArrayParseTable(ParseTable):
+    """
+    Stores the parse table in a array of 16bit ints with all the states
+    packed one after the other in the following structure:
+
+        num_tokens [ token_id atom ]*
+
+    The atom has its highest bit set if it's a reduce action.
+    """
```

Comment:
Mixed feelings with this, it reduces memory a bit but it's a bit slower even with a cache. I don't think this could be much more optimized without going the cython route or similar. Maybe on PyPy this approach works better, it should be able to JIT the binary search loop easily.